src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -15,6 +15,46 @@ static int win_init_sep(MPIR_Win * win);
 static int win_init_stx(MPIR_Win * win);
 static int win_init_global(MPIR_Win * win);
 static int win_init(MPIR_Win * win);
+static uint64_t win_get_global_mr_key(MPIR_Comm * comm_ptr);
+
+
+static uint64_t win_get_global_mr_key(MPIR_Comm * comm_ptr)
```

Comment:
What provider is this in support of?

Reply:
I thought we had agreed this was wrong:
```
When the MPIDI_OFI_ENABLE_MR_PROV_KEY is by default set to FALSE in psm2
```
and that we were going to set it to TRUE to fix it.

Reply:
> What provider is this in support of?

This supports providers that allow user-defined mr key


> I thought we had agreed this was wrong:
> 
> ```
> When the MPIDI_OFI_ENABLE_MR_PROV_KEY is by default set to FALSE in psm2
> ```
> 
> and that we were going to set it to TRUE to fix it.

Sorry, I discussed with Min after that; we think the issue is not `MPIDI_OFI_ENABLE_MR_PROV_KEY`; instead, PSM2 allows users to provide mr key, but MPICH does not give a consistent key among processes in a window. This is the actual root cause.

Reply:
Sure, that is a solution to the problem. But that problem is because of an incorrect optimization, i.e. discarding the key data structure when it was still needed. Another fix would have been to not discard it. We should be clear in the description of these changes.

Reply:
Sure, I can complement my problem description.
We didn't keep winfo data structure here because if providers allow the user to provide mr key, we can optimize the winfo array out by guaranteeing the same mr key among processes in a window. winfo array is not scalable data structure, we need to avoid it as much as possible unless we really need it (i.e. when we have to use the provider's mr key)

Reply:
By the way, would it be possible to use the `context_id` from the window communicator as the request MR key? It is guaranteed to be unique among the group of processes associated with the window.

Reply:
Noting that ofi netmod pt2pt and AM code calls `fi_mr_reg` using the same key allocator as RMA. In order to use `context_id`, we would need to isolate RMA registrations from the others, probably with a special bit.

Reply:
Reading the commit `e7c15713a6b6798d623de3504ebcd9cf1b697568`, it looks like the old version was a per-comm key allocator which should resolve this bug. @raffenet is the old per-comm allocator similar to the proposed `context_id` + `special bit` solution?

Reply:
Perhaps the `context_id` as prefix is enough to differentiate the two cases. Since windows will always dup the input communicator, there is no change of colliding with `context_id` 0 (MPI_COMM_WORLD), so you can assume a prefix of 0 for all non-window registrations.

Reply:
Correct my comment for the MR registration dynamic window part: we just use the allocator-generated `mr_key` for attached segments and always exchange/store O(P) `mr_keys` at "collective" win_attach, no matter `MPIDI_OFI_ENABLE_MR_PROV_KEY` is on or off.

Reply:
I am not very familiar with the internal structure of key. Since each win always dups the input comm, even using the same comm for two wins, the context_id will be different. So we need to only ensure that a `context_id` is never equal to the allocator-generated key (e.g., by having a few prefix bits and set different values?). Am I understanding correctly?

Reply:
I feel if we use context_id for mr key, key management will become more complex. If users continuously call AM ACC/pt2pt which exceeds the number of mr key context_id can offer, we have to consider expand context_id, so I would suggest we just use the current allocator for all mr key generation instead of context_id.

Reply:
To clarify my point, I am not suggesting using context_id **only** as mr key. What I am suggesting is setting aside bits in the mr key to be used as a "collective" key, and using context_id as that value.

Context ids are collectively allocated values, so I think adding a similar allocation scheme for mr keys would be unnecessary. We just need to figure out the API changes needed to support all types of registrations in MPICH.

Reply:
sounds good, I have some ideas to implement it, will push it soon.

Reply:
Let me know if you want to discuss offline. It should be easier than back and forth on here.

Reply:
got it